### PR TITLE
Some bug fix for load task and schema manage

### PR DESCRIPTION
**Pull request by Linary**

### CODE REVIEW

File: hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java [hunk 1]
```diff
@@ -138,64 +144,71 @@ public LoadTask(LoadOptions options, GraphConnection connection,
...
+                if (noError) {
+                    this.status = LoadStatus.SUCCEED;
+                } else {
+                    this.status = LoadStatus.FAILED;
+                }
+            }
+            this.fileReadLines = this.context().newProgress().totalInputReaded();
+            this.lastDuration += this.context().summary().totalTime();
+            this.currDuration = 0L;
+            // how to save db
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java:181]: remove it

File: hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java [hunk 2]
```diff
@@ -138,64 +144,71 @@ public LoadTask(LoadOptions options, GraphConnection connection,
...
+            this.lock.unlock();
+        }
+    }
+
+    public void stop() {
+        this.context().stopLoading();
+        while (this.status.inRunning()) {
+            try {
+                Thread.sleep(1000);
+            } catch (InterruptedException e) {
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java:192]: rename to ignored

File: hubble-be/src/main/java/com/baidu/hugegraph/handler/LoadTaskExecutor.java [hunk 3]
```diff
@@ -31,8 +31,10 @@
 public class LoadTaskExecutor {
 
     @Async
-    public void execute(LoadTask task) {
+    public void execute(LoadTask task, Runnable callback) {
         log.info("LoadTask is executing run task:{}", task.getId());
         task.run();
+        log.info("Execute callback");
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/handler/LoadTaskExecutor.java:37]: update log to:
log.info("Executing task: {}", task.getId());
log.info("Executed task: {}", task.getId());

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java [hunk 4]
```diff
@@ -112,31 +113,38 @@ public int remove(int id) {
...
-        log.debug("Upload file {} length {}", fileName, srcFile.getSize());
-        FileUploadResult result = new FileUploadResult();
-        result.setName(fileName);
-        result.setSize(srcFile.getSize());
+        log.debug("Uploading file {} length {}", partName, srcFile.getSize());
         try {
             // transferTo should accept absolute path
             srcFile.transferTo(destFile.getAbsoluteFile());
             result.setStatus(FileUploadResult.Status.SUCCESS);
+            log.info("Uploaded file part {}", partName + "-" + index);
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java:144]: "Uploaded file part {}-{}",

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java [hunk 5]
```diff
@@ -160,7 +168,8 @@ public boolean tryMergePartFiles(String dirPath, int total) {
                 // Rename file to dest file
                 FileUtils.moveFile(partFiles[0], newFile);
             } catch (IOException e) {
-                throw new InternalException("load.upload.move-file.failed");
+                log.error(e);
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java:171]: add some message for log

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java [hunk 6]
```diff
@@ -178,19 +187,22 @@ public boolean tryMergePartFiles(String dirPath, int total) {
                     try (InputStream is = new FileInputStream(partFile)) {
                         IOUtils.copy(is, os);
                     } catch (IOException e) {
+                        log.error(e);
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java:190]: ditto

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java [hunk 7]
```diff
@@ -178,19 +187,22 @@ public boolean tryMergePartFiles(String dirPath, int total) {
...
+                log.error(e);
+                throw new InternalException("load.upload.merge-file.failed", e);
             }
         }
         // Delete origin directory
         try {
             FileUtils.forceDelete(dir);
         } catch (IOException e) {
-            throw new InternalException("load.upload.delete-temp-dir.failed");
+            log.error(e);
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java:204]: ditto

File: hubble-be/src/main/java/com/baidu/hugegraph/service/algorithm/AsyncTaskService.java [hunk 8]
```diff
@@ -65,9 +63,7 @@ public Task get(int connId, int id) {
         }
         List<Task> list = client.task().list(status);
         List<Task> result = new ArrayList<>();
-        Iterator<Task> tasks = list.iterator();
-        while (tasks.hasNext()) {
-            Task task = tasks.next();
+        for (Task task : list) {
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/algorithm/AsyncTaskService.java:66]: rename list to tasks

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java [hunk 9]
```diff
@@ -294,19 +295,25 @@ public void pauseAllTasks() {
...
-            task.setDuration(context.summary().totalTime());
-            if (this.update(task) != 1) {
-                throw new InternalException("entity.update.failed", task);
+            task.lock.lock();
+            try {
+                if (task.getStatus().inRunning()) {
+                    LoadContext context = task.context();
+                    task.setFileReadLines(context.newProgress().totalInputReaded());
+                    task.setCurrDuration(context.summary().totalTime());
+                    if (this.update(task) != 1) {
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java:311]: what does 1 mean, prefer boolean or enum

Comment (Linary) [reply hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java:311]: 1 means operate affected rows. it defined by mybatis-plus framework.

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java [hunk 10]
```diff
@@ -178,19 +188,23 @@ public boolean tryMergePartFiles(String dirPath, int total) {
                     try (InputStream is = new FileInputStream(partFile)) {
                         IOUtils.copy(is, os);
                     } catch (IOException e) {
+                        log.error("Failed copy file stream from {} to {}",
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java:191]: Failed to copy

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java [hunk 11]
```diff
@@ -178,19 +188,23 @@ public boolean tryMergePartFiles(String dirPath, int total) {
...
+                                  partFile, newFile, e);
                         throw new InternalException(
-                                  "load.upload.merge-file.failed");
+                                  "load.upload.merge-file.failed", e);
                     }
                 }
             } catch (IOException e) {
-                throw new InternalException("load.upload.merge-file.failed");
+                log.error(e);
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/FileMappingService.java:198]: improve

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java [hunk 12]
```diff
@@ -204,16 +195,20 @@ public LoadTask resume(int taskId) {
...
-        task.setStatus(LoadStatus.RUNNING);
         // Set work mode in incrental mode, load from last breakpoint
-        task.context().options().incrementalMode = true;
-        this.taskExecutor.execute(task);
-
-        if (this.update(task) != 1) {
-            throw new InternalException("entity.update.failed", task);
+        task.getOptions().incrementalMode = true;
+        task.restoreContext();
+        task.lock.lock();
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java:201]: prefer to add method task.lock()/unlock()

File: hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java [hunk 13]
```diff
@@ -55,14 +54,15 @@
...
 @TableName(value = "load_task", autoResultMap = true)
-@JsonPropertyOrder({"id", "conn_id", "task_id" , "file_id", "file_name", "vertices",
-                    "edges", "load_rate", "load_progress", "file_total_lines",
-                    "file_read_lines", "status", "duration", "create_time"})
 public class LoadTask implements Runnable {
 
     @TableField(exist = false)
     @JsonIgnore
-    private transient HugeGraphLoader loader;
+    public transient final Lock lock = new ReentrantLock();
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java:61]: set private

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/load/FileUploadController.java [hunk 14]
```diff
@@ -69,15 +74,31 @@
...
+    public FileUploadController() {
+        this.uploadingTokenLocks = new ConcurrentHashMap<>();
+    }
+
+    @GetMapping("token")
+    public Response fileToken(@PathVariable("connId") int connId,
+                              @PathVariable("jobId") int jobId,
+                              @RequestParam("name") String fileName) {
+        String token = this.service.generateFileToken(fileName);
+        this.uploadingTokenLocks.put(token, new ReentrantReadWriteLock());
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/controller/load/FileUploadController.java:88]: check exist

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java [hunk 15]
```diff
@@ -294,19 +295,25 @@ public void pauseAllTasks() {
...
     @Async
-    @Scheduled(fixedDelay = 5 * 1000)
+    @Scheduled(fixedDelay = 1 * 1000)
     @Transactional(isolation = Isolation.READ_COMMITTED)
     public void updateLoadTaskProgress() {
-        Map<Integer, LoadTask> taskContainer = this.getTaskContainer();
-        Iterator<Map.Entry<Integer, LoadTask>> iter = taskContainer.entrySet()
-                                                                   .iterator();
+        Iterator<Map.Entry<Integer, LoadTask>> iter;
+        iter = this.taskContainer.entrySet().iterator();
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java:302]: runningTasks

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java [hunk 16]
```diff
@@ -241,10 +240,10 @@ public LoadTask retry(int taskId) {
         Ex.check(task.getStatus() == LoadStatus.FAILED ||
                  task.getStatus() == LoadStatus.STOPPED,
                  "Can only retry the FAILED or STOPPED task");
+        task.lock();
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java:243]: move line 245~246 to try block

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/load/FileMappingController.java [hunk 17]
```diff
@@ -286,6 +291,17 @@ public void loadParameter(@RequestBody LoadParameter newEntity) {
         }
     }
 
+    @PutMapping("finish")
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/controller/load/FileMappingController.java:294]: prefer rename to another name, like ready-loading

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/load/FileUploadController.java [hunk 18]
```diff
@@ -240,6 +247,17 @@ public Boolean delete(@PathVariable("connId") int connId,
         }
     }
 
+    @PutMapping("finish")
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/controller/load/FileUploadController.java:250]: ditto

File: hubble-be/src/main/java/com/baidu/hugegraph/options/HubbleOptions.java [hunk 19]
```diff
@@ -192,6 +192,16 @@ public static synchronized HubbleOptions instance() {
                     10 * Bytes.GB
             );
 
+    public static final ConfigOption<Long> UPLOAD_FILE_MAX_TIME_CONSUMING =
+            new ConfigOption<>(
+                    "upload_file.max_uploading_time",
+                    "The maximum allowable uploading time(second) for file " +
+                    "uploads, the uploaded parts will be cleared if exceed " +
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/options/HubbleOptions.java:199]: uploaded file parts

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java [hunk 20]
```diff
@@ -154,48 +146,47 @@ public int update(LoadTask entity) {
...
         if (this.save(task) != 1) {
             throw new InternalException("entity.insert.failed", task);
         }
-        this.taskExecutor.execute(task);
-
-        if (this.update(task) != 1) {
-            throw new InternalException("entity.update.failed", task);
-        }
+        // executed in other threads
+        this.taskExecutor.execute(task, () -> this.update(task));
```
[the remark this answers is not in the document]
Comment (javeme) [reply hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java:170]: should handle exception if the update is failed?

File: hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java [hunk 21]
```diff
@@ -143,59 +135,60 @@ public int taskCountByJob(int jobId) {
...
-            throw new InternalException("entity.insert.failed", task);
-        }
-        this.taskExecutor.execute(task);
-
-        if (this.update(task) != 1) {
-            throw new InternalException("entity.update.failed", task);
-        }
+        this.save(task);
+        // executed in other threads
+        this.taskExecutor.execute(task, () -> this.update(task));
```
Comment (javeme) [inline hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java:170]: should handle exception if the update is failed?

Comment (Linary) [reply hubble-be/src/main/java/com/baidu/hugegraph/service/load/LoadTaskService.java:170]: no need, it doesn't affect other tasks to use the thread pool

**Review by javeme — APPROVED** [review]

File: hubble-be/src/main/java/com/baidu/hugegraph/controller/load/FileMappingController.java [hunk 22]
```diff
@@ -130,18 +132,14 @@ public FileMapping fileSetting(@PathVariable("id") int id,
         if (mapping == null) {
             throw new ExternalException("load.file-mapping.not-exist.id", id);
         }
-        // unescape \\t to \t
-        newEntity.unescapeDelimiterIfNeeded();
+        // change format to TEXT if needed
```
Comment (zhoney) [inline hubble-be/src/main/java/com/baidu/hugegraph/controller/load/FileMappingController.java:135]: // Change format to TEXT if needed

File: hubble-be/src/main/java/com/baidu/hugegraph/entity/load/FileMapping.java [hunk 23]
```diff
@@ -157,4 +151,26 @@ public EdgeMapping getEdgeMapping(String emId) {
         }
         return null;
     }
+
+    @JsonIgnore
+    public Set<String> getVertexMappingLabels() {
+        if (this.getVertexMappings() != null) {
```
Comment (zhoney) [inline hubble-be/src/main/java/com/baidu/hugegraph/entity/load/FileMapping.java:157]:
```
if (this.getVertexMappings() != null) {
    return new HashSet<>();
}
return this.getVertexMappings().stream()
           .map(ElementMapping::getLabel)
           .collect(Collectors.toSet());
```

File: hubble-be/src/main/java/com/baidu/hugegraph/entity/load/FileMapping.java [hunk 24]
```diff
@@ -157,4 +151,26 @@ public EdgeMapping getEdgeMapping(String emId) {
...
+                       .map(ElementMapping::getLabel)
+                       .collect(Collectors.toSet());
+        } else {
+            return new HashSet<>();
+        }
+    }
+
+    @JsonIgnore
+    public Set<String> getEdgeMappingLabels() {
+        if (this.getEdgeMappings() != null) {
```
Comment (zhoney) [inline hubble-be/src/main/java/com/baidu/hugegraph/entity/load/FileMapping.java:168]: ditto

File: hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java [hunk 25]
```diff
@@ -132,70 +142,87 @@
...
-        // Pay attention to whether the user stops actively or
-        // the program stops by itself
-        if (this.status.inRunning()) {
-            if (succeed) {
-                this.status = LoadStatus.SUCCEED;
-            } else {
-                this.status = LoadStatus.FAILED;
+        this.lock.lock();
+        try {
+            // Pay attention to whether the user stops actively or
```
Comment (zhoney) [inline hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java:174]: /*
 */

File: hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java [hunk 26]
```diff
@@ -205,14 +232,26 @@ public int getLoadProgress() {
...
+
     @JsonProperty("load_rate")
     public String getLoadRate() {
+        long readLines = this.fileReadLines;
+        long duration = this.getDuration();
         float rate;
-        if (this.fileReadLines == null || this.duration == null ||
-            this.duration == 0L) {
+        if (readLines == 0L || duration == 0L) {
             rate = 0;
```
Comment (zhoney) [inline hubble-be/src/main/java/com/baidu/hugegraph/entity/load/LoadTask.java:252]: rate = 0.0F

**Review by zhoney — CHANGES_REQUESTED** [review]

**Review by zhoney — APPROVED** [review]

**Review by javeme — APPROVED** [review]

**Pull request merged**